crates/bevy_reflect/src/func/args/list.rs

```diff
@@ -0,0 +1,81 @@
+use crate::func::args::Arg;
+use crate::Reflect;
+
+/// A list of arguments that can be passed to a dynamic [`Function`].
+///
+/// # Example
+///
+/// ```
+/// # use bevy_reflect::func::{Arg, ArgList};
+/// let foo = 123;
+/// let bar = 456;
+/// let mut baz = 789;
+/// let args = ArgList::new()
+///   // Push an owned argument
+///   .push_owned(foo)
+///    // Push an owned and boxed argument
+///   .push_boxed(Box::new(foo))
+///   // Push a reference argument
+///   .push_ref(&bar)
+///   // Push a mutable reference argument
+///   .push_mut(&mut baz)
+///   // Push a manually constructed argument
+///   .push(Arg::Ref(&3.14));
+/// ```
+///
+/// [`Function`]: crate::func::Function
+#[derive(Default, Debug)]
+pub struct ArgList<'a>(Vec<Arg<'a>>);
```

Comment:
Pretty much. In one iteration I had originally used an enum in an effort to avoid `Vec` allocations:

```rust
enum ArgList<'a> {
  Arg0,
  Arg1(Arg<'a>),
  Arg2(Arg<'a>, Arg<'a>),
  Arg3(Arg<'a>, Arg<'a>, Arg<'a>),
  // ...
  Variadic(Vec<Arg<'a>>)
}
```

If we did that, we potentially could maintain some degree of lifetimes by just adding a bunch of lifetimes to `ArgList`:

```rust
enum ArgList<'a, 'b, 'c, 'default: 'a + 'b + 'c> {
  Arg0,
  Arg1(Arg<'a>),
  Arg2(Arg<'a>, Arg<'b>),
  Arg3(Arg<'a>, Arg<'b>, Arg<'c>),
  // ...
  Variadic(Vec<Arg<'default>>)
}
```

The output would still only be able to tie itself to the first argument's lifetime, but doing this would mean that we shouldn't need to shrink lifetimes down (most of the time).

So that might be something to explore. 

Thoughts?

Reply:
I think the enum `ArgList` might work; and it wouldn't even require the `Variadic` variant no?
So every new enum added would update from `ArgN` to `ArgN+1`.

Might be something to explore, but I think it's also ok to merge the functionality as is and explore this in a future PR

Reply:
> I think the enum `ArgList` might work; and it wouldn't even require the `Variadic` variant no?

The `Variadic` would be needed so users could supply more arguments than we have variants for. While `IntoFunction` supports a limited number of arguments, users are still able to construct their `Function` manually, with as many arguments as they want.

> Might be something to explore, but I think it's also ok to merge the functionality as is and explore this in a future PR

Yeah that might be a good idea. Any breakage would be minimal and really only apply to cases where the lifetime of `ArgList` is explicitly set by a user.

Reply:
Yeah definitely going to save this for a followup PR!

Reply:
Yeah, please leave this to followup.